CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityLinkInventoryBase.h

```diff
@@ -0,0 +1,233 @@
+	class FPPLinkListBase : public std::vector<T*> {
+	public:
+		typedef typename std::vector<T*>::iterator FPPLinkListIterator_t;
+		virtual ~FPPLinkListBase() {}
+	};
+
+	class FOCALPOINTPLASTICITY_EXPORT FPPLinkID {
+	public:
+		long id0;
+		long id1;
+
+		FPPLinkID() : id0(0), id1(0) {};
+		FPPLinkID(long _id0, long _id1)
+		{
+			if (_id0 < _id1) {
+				id0 = _id0;
+				id1 = _id1;
+			}
+			else {
+				id1 = _id0;
+				id0 = _id1;
+			}
+		}
+		~FPPLinkID() {}
+		bool operator < (const FPPLinkID & _rhs) const { return id0 < _rhs.id0 || (id0 == _rhs.id0 && id1 < _rhs.id1); }
+		bool operator == (const FPPLinkID & _rhs) const { return id0 == _rhs.id0 && id1 == _rhs.id1; }
+		bool operator != (const FPPLinkID & _rhs) const { return !(operator==(_rhs)); }
+	};
+
+	template <class LinkType> class FPPLinkInventoryBase;
+
+	template <class LinkType>
+	class FOCALPOINTPLASTICITY_EXPORT FPPLinkInventoryTracker {
+	public:
+
+		FPPLinkInventoryTracker() {};
+		~FPPLinkInventoryTracker() {};
+		
+		FPPLinkInventoryBase<LinkType> linkInv;
+	};
+
+
+	template <class LinkType>
+	class FOCALPOINTPLASTICITY_EXPORT FPPLinkInventoryBase {
+
+	public:
+
+		typedef FPPLinkListBase<LinkType> FPPLinkList;
+		typedef FPPLinkInventoryBase<LinkType> FPPInventory_t;
+
+		typedef std::map<const FPPLinkID, LinkType*> linkInventory_t;
+		typedef typename linkInventory_t::iterator linkInventoryItr_t;
+		typedef std::pair<const FPPLinkID, LinkType*> linkInventoryPair_t;
+
+	protected:
+
+		Potts3D* potts;
+		BasicClassAccessor<FocalPointPlasticityTracker>* focalPointPlasticityTrackerAccessor;
+		BasicClassAccessor<FPPLinkInventoryTracker<LinkType> >* cellLinkInventoryTrackerAccessor;
+
+		linkInventory_t linkInventory;
+
+		LinkType* getLinkById(const FPPLinkID _id) {
+			linkInventoryItr_t itr = linkInventory.find(_id);
+			if (itr != linkInventory.end()) return itr->second;
+			return (LinkType*)(0);
+		}
+		const FPPLinkID getLinkId(LinkType* _link) { return FPPLinkID(_link->getId0(), _link->getId1()); }
+
+		// Add a link without any additional internal work
+		void addLinkNoChain(LinkType* _link) {
+			linkInventory.insert(linkInventoryPair_t(getLinkId(_link), _link));
+		}
+		// Add tracker data to cells
+		void addTrackerData(LinkType* _link) {
+			CellG* cell0 = _link->getObj0();
+			CellG* cell1 = _link->getObj1();
+			if (cell0) {
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell0);
+				getFPPTrackerDataSet(cell0).insert(fppd);
+			}
+			if (cell1) {
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell1);
+				getFPPTrackerDataSet(cell1).insert(fppd);
+			}
+		}
+		// Remove a link without any additional internal work
+		void removeLinkNoChain(LinkType* _link) {
+			const FPPLinkID linkId = getLinkId(_link);
+			linkInventoryItr_t itr = linkInventory.find(linkId);
+			if (itr != linkInventory.end()) linkInventory.erase(linkId);
+		}
+		// Remove tracker data from cells
+		void removeTrackerData(LinkType* _link) {
```

Comment:
Since we are overhauling everything I woudl get rid of TrackerData from cells. This means that no cell should have ANY reference to FPPTrackerData. Instead we should write a code that "simulates existence" of FPPTrackerData and I think you are doing it already. I want to make sure that we move away from this implicit link definition that we currently have

Reply:
Simulating the existence of FPPTrackerData is exactly what this is intended to do. Currently the reference to FPPTrackerData is still defined for each cell, though the data is now managed by the link inventories. To my knowledge, the only thing this now supports is the Python iterators for the data list. However, we can generate the data for those iterators from the link inventories. I'll make this change and remove FPPTrackerData from the cells altogether before we merge.

Reply:
Implemented in https://github.com/CompuCell3D/CompuCell3D/pull/361/commits/4ce8f702eb17e2ca303319d02b23df9676e47ed4